### PR TITLE
Improve GitLab CI

### PR DESCRIPTION
**Pull request by Robbepop**
Improvements to the `scripts/check-workspace.sh` as well as general fixes throughout the whole crate. This is a follow-up on @koushiro 's work but more elaborate. Fixes on the `.gitlab.yml` are in preparation so that our GitLab CI will finally be fixed. Also we might think about removing Travis CI already in this PR because it holds us back.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -90,13 +90,16 @@ check-std:
   stage:                           check-workspace
   <<:                              *docker-env
   script:
-    - cargo check --verbose --all-features --release --all
+    - all_crates=("core" "alloc" "utils" "lang2" "lang2/macro" "cli")
```
Comment (Robbepop) [inline .gitlab-ci.yml]: Can we somehow fix this code dupe? Updating this would be really daunting.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -33,6 +33,7 @@ variables:
     - rustup show
     - bash --version
     # global RUSTFLAGS overrides the linker args so this way is better to pass the flags
```
Comment (Robbepop) [inline .gitlab-ci.yml]: Comment should be moved 1 line down.

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -52,73 +53,131 @@ variables:
...
 
-cargo-check-nf:
+check-wasm:
   stage:                           check-workspace
   <<:                              *docker-env
   script:
-  - cargo check --verbose --all --no-default-features
+    - wasm_crates=("core" "alloc" "utils" "lang2" "lang2/macro")
+    - for crate in "${wasm_crates[@]}"; do
+        cargo check --verbose --no-default-features --release --target wasm32-unknown-unknown --manifest-path ${crate}/Cargo.toml;
```
Comment (Robbepop) [inline .gitlab-ci.yml]:
```suggestion
        cargo check --verbose --no-default-features --target wasm32-unknown-unknown --manifest-path ${crate}/Cargo.toml;
```

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -52,73 +53,131 @@ variables:
...
 
-cargo-check-af:
+check-std:
   stage:                           check-workspace
   <<:                              *docker-env
   script:
-  - cargo check --verbose --all --all-features
+    - all_crates=("core" "alloc" "utils" "lang2" "lang2/macro" "cli")
+    - for crate in "${wasm_crates[@]}"; do
+        cargo check --verbose --all-features --release;
```
Comment (Robbepop) [inline .gitlab-ci.yml]:
```suggestion
        cargo check --verbose --all-features;
```

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -52,73 +53,131 @@ variables:
...
+clippy-std:
   stage:                           check-workspace
   <<:                              *docker-env
   script:
-  - cargo test --verbose --all --all-features
+    - all_crates=("core" "alloc" "utils" "lang2" "lang2/macro" "cli")
+    - for crate in "${wasm_crates[@]}"; do
+        cargo clippy --verbose --all-features -- -D warnings;
+      done
+  allow_failure:                   true
```
Comment (Robbepop) [inline .gitlab-ci.yml]: Why allow failure?

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -52,73 +53,131 @@ variables:
...
+clippy-wasm:
   stage:                           check-workspace
   <<:                              *docker-env
   script:
-  - cargo build --verbose --all --no-default-features --release --target=wasm32-unknown-unknown
+    - wasm_crates=("core" "alloc" "utils" "lang2" "lang2/macro")
+    - for crate in "${wasm_crates[@]}"; do
+        cargo clippy --verbose --manifest-path ${crate}/Cargo.toml --no-default-features -- -D warnings;
+      done
+  allow_failure:                   true
```
Comment (Robbepop) [inline .gitlab-ci.yml]: Why allow failure?

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -52,73 +53,131 @@ variables:
...
 #### stage:                        examples
 
-examples-wasm:
+examples-test:
   stage:                           examples
   <<:                              *docker-env
   script:
-    - for example in $(ls examples/lang); do
-        cargo build --release --no-default-features --target=wasm32-unknown-unknown --verbose --manifest-path examples/lang/$example/Cargo.toml;
+    - for DIR in examples/lang2/*; do
```
Comment (Robbepop) [inline .gitlab-ci.yml]: I think the name `example` here was totally fine.

File: .gitlab-ci.yml [hunk 8]
```diff
@@ -52,73 +53,131 @@ variables:
...
+        cargo clippy --verbose --manifest-path ${DIR}/Cargo.toml --no-default-features -- -D warnings;
+      done
+  allow_failure:                   true
+
+examples-contract-build:
+  stage:                           examples
+  <<:                              *docker-env
+  script:
+    - for DIR in examples/lang2/*; do 
+        cd $DIR && pwd && cargo contract build;
```
Comment (Robbepop) [inline .gitlab-ci.yml]: I'd go with a
```
pushd $example && cargo contract build && popd;
```

File: .gitlab-ci.yml (hunk 1 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: I'd loop all dirs, but there you have `model` and `lang` which do not need tests.
Will think how to make it better.

File: .gitlab-ci.yml (hunk 2 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: actually `mkdir` and `touch` fix the previous problem we had with writing to `CARGO_HOME` for the first time, but ok.

File: .gitlab-ci.yml [hunk 9]
```diff
@@ -171,13 +171,19 @@ examples-contract-build:
   <<:                              *docker-env
   script:
     - for DIR in examples/lang2/*; do 
-        cd $DIR && pwd && cargo contract build; 
+        cd $DIR && 
+        pwd &&
```
Comment (Robbepop) [inline .gitlab-ci.yml]: Why `pwd` needed?

File: .gitlab-ci.yml [hunk 10]
```diff
@@ -52,73 +53,137 @@ variables:
...
-    - for example in $(ls examples/lang); do
-        cargo test --verbose --manifest-path examples/lang/$example/Cargo.toml;
+    - for DIR in examples/lang2/*; do 
+        cargo fmt --verbose --manifest-path ${DIR}/Cargo.toml -- --check;
       done
-    - sccache -s
+  allow_failure:                   true
 
-examples-abi:
+examples-clippy-all-features:
```
Comment (Robbepop) [inline .gitlab-ci.yml]:
```suggestion
examples-clippy-std:
```

File: .gitlab-ci.yml [hunk 11]
```diff
@@ -52,73 +53,137 @@ variables:
...
   <<:                              *docker-env
   script:
-    - for example in $(ls examples/lang); do
-        cargo run --package abi-gen --manifest-path examples/lang/$example/Cargo.toml;
+    - for DIR in examples/lang2/*; do 
+        cargo clippy --verbose --manifest-path ${DIR}/Cargo.toml --all-features -- -D warnings;
+      done
+  allow_failure:                   true
+
+examples-clippy-no-default-features:
```
Comment (Robbepop) [inline .gitlab-ci.yml]:
```suggestion
examples-clippy-wasm:
```

File: .gitlab-ci.yml (hunk 5 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: will fix

File: .gitlab-ci.yml (hunk 8 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: That's the same

File: .gitlab-ci.yml (hunk 9 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: for you to know where you run tests!

File: .gitlab-ci.yml (hunk 8 above)
Comment (Robbepop) [reply .gitlab-ci.yml]: It is a more robust solution in general.

File: .gitlab-ci.yml [hunk 12]
```diff
@@ -16,6 +16,8 @@ variables:
   CARGO_INCREMENTAL:               0
   CI_SERVER_NAME:                  "GitLab CI"
   REGISTRY:                        registry.parity.io/parity/infrastructure/scripts
+  ALL_CRATES:                      "core alloc utils lang2 lang2/macro cli"
+  WASM_CRATES:                     "core alloc utils lang2 lang2/macro"
```
Comment (Robbepop) [inline .gitlab-ci.yml]: Could you maybe even reuse `WASM_CRATES` in `ALL_CRATES` by using `$WASM_CRATES`?

File: .gitlab-ci.yml (hunk 1 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: Created `${ALL_CRATES}` and `${WASM_CRATES}` global variables with lists of the crates.

File: .gitlab-ci.yml (hunk 12 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: GitLab can't set lists to variables. This way it's explanatory enough.

File: .gitlab-ci.yml (hunk 7 above)
Comment (TriplEight) [reply .gitlab-ci.yml]: ok

File: .gitlab-ci.yml [hunk 13]
```diff
@@ -52,73 +55,121 @@ variables:
...
-    - sccache -s
 
-examples-test:
+examples-fmt:
   stage:                           examples
   <<:                              *docker-env
   script:
-    - for example in $(ls examples/lang); do
-        cargo test --verbose --manifest-path examples/lang/$example/Cargo.toml;
+    - for DIR in examples/lang2/*; do
```
Comment (Robbepop) [inline .gitlab-ci.yml]:
```suggestion
    - for example in examples/lang2/*; do 
```

File: .gitlab-ci.yml [hunk 14]
```diff
@@ -52,73 +55,121 @@ variables:
...
 
-examples-test:
+examples-fmt:
   stage:                           examples
   <<:                              *docker-env
   script:
-    - for example in $(ls examples/lang); do
-        cargo test --verbose --manifest-path examples/lang/$example/Cargo.toml;
+    - for DIR in examples/lang2/*; do 
+        cargo fmt --verbose --manifest-path ${DIR}/Cargo.toml -- --check;
```
Comment (Robbepop) [inline .gitlab-ci.yml]:
```suggestion
        cargo fmt --verbose --manifest-path ${example}/Cargo.toml -- --check;
```

File: .gitlab-ci.yml [hunk 15]
```diff
@@ -52,73 +55,121 @@ variables:
...
-    - sccache -s
 
-examples-abi:
+examples-clippy-std:
   stage:                           examples
   <<:                              *docker-env
   script:
-    - for example in $(ls examples/lang); do
-        cargo run --package abi-gen --manifest-path examples/lang/$example/Cargo.toml;
+    - for DIR in examples/lang2/*; do
```
Comment (Robbepop) [inline .gitlab-ci.yml]:
```suggestion
    - for example in examples/lang2/*; do 
```

File: .gitlab-ci.yml [hunk 16]
```diff
@@ -52,73 +55,121 @@ variables:
...
 
-examples-abi:
+examples-clippy-std:
   stage:                           examples
   <<:                              *docker-env
   script:
-    - for example in $(ls examples/lang); do
-        cargo run --package abi-gen --manifest-path examples/lang/$example/Cargo.toml;
+    - for DIR in examples/lang2/*; do 
+        cargo clippy --verbose --manifest-path ${DIR}/Cargo.toml --all-features -- -D warnings;
```
Comment (Robbepop) [inline .gitlab-ci.yml]:
```suggestion
        cargo clippy --verbose --manifest-path ${example}/Cargo.toml --all-features -- -D warnings;
```

File: .gitlab-ci.yml [hunk 17]
```diff
@@ -52,73 +55,121 @@ variables:
...
+      done
+
+examples-clippy-wasm:
+  stage:                           examples
+  <<:                              *docker-env
+  script:
+    - for DIR in examples/lang2/*; do
```
Comment (Robbepop) [inline .gitlab-ci.yml]:
```suggestion
    - for example in examples/lang2/*; do 
```

File: .gitlab-ci.yml [hunk 18]
```diff
@@ -52,73 +55,121 @@ variables:
...
+    - for DIR in examples/lang2/*; do 
+        cargo clippy --verbose --manifest-path ${DIR}/Cargo.toml --all-features -- -D warnings;
+      done
+
+examples-clippy-wasm:
+  stage:                           examples
+  <<:                              *docker-env
+  script:
+    - for DIR in examples/lang2/*; do 
+        cargo clippy --verbose --manifest-path ${DIR}/Cargo.toml --no-default-features -- -D warnings;
```
Comment (Robbepop) [inline .gitlab-ci.yml]:
```suggestion
        cargo clippy --verbose --manifest-path ${example}/Cargo.toml --no-default-features -- -D warnings;
```

File: cli/src/cmd/abi.rs [hunk 19]
```diff
@@ -26,13 +24,17 @@ use std::path::PathBuf;
...
-        "--no-default-features",
-        "--verbose",
-    ], dir)?;
+    super::exec_cargo(
+        "run",
+        &[
+            "--package",
+            "abi-gen",
+            "--release",
+            "--no-default-features",
```
Comment (Robbepop) [inline cli/src/cmd/abi.rs]:
```suggestion
            // "--no-default-features", // Breaks builds for MacOS (linker errors), we should investigate this issue asap!
```

**Pull request merged**